src/System.Private.CoreLib/shared/System/Resources/ManifestBasedResourceGroveler.cs

```diff
@@ -465,5 +462,40 @@ private void HandleResourceStreamMissing(string fileName)
             resName += fileName;
             throw new MissingManifestResourceException(SR.Format(SR.MissingManifestResource_NoNeutralAsm, resName, _mediator.MainAssembly.GetName().Name));
         }
+
+        private static bool GetNeutralResourcesLanguageAttribute(Assembly assembly, out string cultureName, out UltimateResourceFallbackLocation fallbackLocation)
+        {
+            foreach (CustomAttributeData cad in assembly.GetCustomAttributesData())
```

Comment:
This does a lot more work and can fail for many more reasons than the runtime impl. Will need to look at perf&compat before merging.

Reply:
It is very similar to the CoreRT implementation (with added `Assert`s). I didn't initially adopt it because of concerns about caching of the values in the runtime implementation and possible performance implications.

At some point I used this readable code (at the expense of performance; although I never benchmarked it):
```c#
            var nrlAttribute = assembly.GetCustomAttribute<NeutralResourcesLanguageAttribute>();
            if (nrlAttribute != null)
            {
                fallbackLocation = (short)nrlAttribute.Location;
                cultureName = nrlAttribute.CultureName;
                return true;
            }
```

Reply:
The caching was not used in this code path. Is GetCustomAttributesData call your concern?

Reply:
> The caching was not used in this code path.

I wasn't entirely sure about it. The code path didn't cache the value but it would use one if it was cached already. I didn't analyze the code further whether that actually happens or not.

Reply:
> Is GetCustomAttributesData call your concern?

Yes, GetCustomAttributesData will load types of all custom attributes of the assembly via reflection. It both takes time and it can fail for many different reasons. For example, consider something like this:

```
using System;
using System.Reflection;
using System.Runtime.InteropServices;

[assembly: AttributeInAssemblyThatDoesNotExist]

class My
{
    static void Main()
    {
        typeof(My).Assembly.GetCustomAttributesData();
    }
}
```

This will throw FileNotFoundException.

Reply:
> `GetCustomAttribute<NeutralResourcesLanguageAttribute>`

Something like this would look better to me - if it actually did not load the types for all custom attributes underneath. It is not the case today unfortunately - it first loads everything and then it throws away what it does not like.

Reply:
I think `GetCustomAttribute<NeutralResourcesLanguageAttribute>` is not suitable here, it just creates full attribute instances and does all sort of redundant checks and allocations.

The code is fine for CoreRT and Mono should I just ifdef it out for CoreCLR and keep the internal call for CoreCLR?

Reply:
> it just creates full attribute instances and does all sort of redundant checks and allocations.

It doesn't necessarily have to create instances as long as the overload with `inherit: false` is used throughout all the custom attribute methods (not sure that's the case now). In that case it can filter the attributes based on the metadata token and instantiate only the necessary one

I had the same concern as you initially did, but Jan's answer suggests that `Assembly.GetCustomAttributesData` instantiates the attributes too. I assumed it doesn't and that's why your code made sense to me.

Reply:
> it just creates full attribute instances and does all sort of redundant checks and allocations.

The perf numbers say otherwise. On current .NET Core 3.0:

```
// Takes 516ms
for (int i = 0; i < 100000; i++)
    typeof(object).Assembly.GetCustomAttribute<NeutralResourcesLanguageAttribute>();
```

```
// Takes 4766ms
for (int i = 0; i < 100000; i++)
    GetNeutralResourcesLanguageAttribute(typeof(object).Assembly, out _, out _);
```